docs: add guide to standalone components preview (#46058)
angular/angular: aio/content/guide/standalone-preview.md

# Getting started with standalone components

In v14 and higher, **standalone components** provide a simplified way to build Angular applications. Standalone components, directives, and pipes aim to streamline the authoring experience by reducing the need for `NgModule`s. Existing applications can optionally and incrementally adopt the new standalone style without any breaking changes.

## Creating standalone components

### The `standalone` flag and component `imports`

Components, directives, and pipes can now be marked as `standalone: true`. Angular classes marked as standalone do not need to be declared in an `NgModule` (the Angular compiler will report an error if you try).

> Standalone components specify their dependencies directly instead of getting them through `NgModules`. For example, if `ImageGridComponent` is a standalone component, it can be imported directly into another standalone component `PhotoGalleryComponent`:

Reviewer: This reads slightly backwards for me: the first sentence is from the perspective of a standalone component that can import dependencies, but then the example approaches it from `ImageGridComponent` which is the component being imported. I think this becomes clearer if the example would be approached starting with `PhotoGalleryComponent`.

```ts
@Component({
  standalone: true,
  selector: 'photo-gallery',
  imports: [ImageGridComponent],
  template: `
    ... <image-grid [images]="imageList"></image-grid>
  `,
})
export class PhotoGalleryComponent {
  // component logic
}
```

> `imports` can also be used to reference standalone directives and pipes. In this way, standalone components can be written without the need to create `NgModule` to manage template dependencies.

Reviewer 2: ```suggestion
`imports` can also be used to reference standalone directives and pipes. In this way, standalone components can be written without the need to create an `NgModule` to manage template dependencies.
```

### Using existing NgModules in a standalone component

> When writing a standalone component, you may want to use other components, directives, or pipes in a template. Some of those dependencies might not be marked as standalone, but instead declared and exported by an existing `NgModule`. In this case, you can import the `NgModule` directly into a standalone component:

Reviewer 3: ```suggestion
When writing a standalone component, you may want to use other components, directives, or pipes in the component's template. Some of those dependencies might not be marked as standalone, but instead declared and exported by an existing `NgModule`. In this case, you can import the `NgModule` directly into the standalone component:
```

```ts
@Component({
  standalone: true,
  selector: 'photo-gallery',
  // an existing module is imported directly into a standalone component
  imports: [MatButtonModule],
  template: `
    ...
    <button mat-button>Next Page</button>
  `,
})
export class PhotoGalleryComponent {
  // logic
}
```

> Writing standalone components doesn’t prevent you from using existing `NgModule`-based libraries or dependencies in your template. Standalone components can take full advantage of the existing ecosystem of Angular libraries.

Reviewer 3: I think it would be nice if this also mentioned that you can import the `NgModule`s of your existing project into new standalone components (it is pretty obvious, but may be still worth mentioning?)

Author: We've got the whole section `Using existing NgModules in a standalone component` about it - do you feel like this is not sufficient?

Reviewer 3: nono I think it is totally sufficient and clear :slightly_smiling_face: 

I just was pointing out that this sentence looks to me very library specific (just saying that maybe an extra comment on existing codebases would not hurt :thinking: :shrug:)

## Using standalone components in NgModule-based applications

> Standalone components can also be imported into existing NgModules-based contexts. This allows existing applications (which are using NgModules today) to incrementally adopt the new, standalone style of component.

Reviewer 3: ```suggestion
Standalone components can also be imported into existing `NgModule`s-based contexts. This allows existing applications (which are using `NgModule`s today) to incrementally adopt the new, standalone style of component.
```
everywhere else you are code-quoting the term

You can import a standalone component (or directive, or pipe) just like you would an `NgModule` - using `NgModule.imports`:

```ts
@NgModule({
  declarations: [AlbumComponent],
  exports: [AlbumComponent], 
  imports: [PhotoGalleryComponent],
})
export class AlbumModule {}
```

## Bootstrapping an application using a standalone component

> An Angular application can be bootstrapped without any `NgModule` by using a standalone component as the application's root component. This is done using the  `bootstrapApplication` API:

Reviewer 3: ```suggestion
An Angular application can be bootstrapped without any `NgModule` by using a standalone component as the application's root component. This is done using the `bootstrapApplication` API:
```

    {path: 'teams', component: AdminTeamsCmp},
  ],
}];
```

Note the use of an empty-path route to host `providers` that are shared among all the child routes.

## Advanced topics

This section goes into more details that are relevant only to more advanced usage patterns. You can safely skip this section when learning about standalone components, directives, and pipes for the first time. 

### Standalone components for library authors

Standalone components, directives, and pipes can be exported from NgModules that import them:

```ts
@NgModule({
  imports: [ImageCarouselComponent, ImageSlideComponent],
  exports: [ImageCarouselComponent, ImageSlideComponent],
})
export class CarouselModule {}
```

This pattern is useful for Angular libraries that publish a set of cooperating directives. In the above example, both the `ImageCarouselComponent` and `ImageSlideComponent` need to be present in a template to build up one logical “carousel widget”. 

As an alternative to publishing a `NgModule`, library authors might want to export an array of  cooperating directives:

```ts
export CAROUSEL_DIRECTIVES = [ImageCarouselComponent, ImageSlideComponent] as const;
```

> Such an array could be imported by applications using `NgModule`s and added to the `@NgModule.declarations`. Please note the presence of the TypeScript’s `as const` construct: it gives Angular compiler additional information required for proper compilation and is a recommended practice (as it makes the exported array immutable from the TypeScript point of view).

Reviewer: This refers to `NgModule.declarations`, but the above example of including `ImageCarouselComponent` and `ImageSlideComponent` in `NgModule.imports` implies that those components are standalone and _cannot_ be included in `NgModule.declarations`. Should this say `NgModule.imports`/`Component.imports` instead?

Author: Yep, should be `@NgModule.imports` - great catch, thnx @JoostK !

Reviewer: It appears that this wasn't updated in the latest push, did the change get lost somehow?

> Such an array could be imported by applications using `NgModule`s and added to the `@NgModule.imports`. Please note the presence of the TypeScript’s `as const` construct: it gives Angular compiler additional information required for proper compilation and is a recommended practice (as it makes the exported array immutable from the TypeScript point of view).

Reviewer 3: ```suggestion
Such an array could be imported by applications using `NgModule`s and added to the `@NgModule.imports`. Please note the presence of the TypeScript’s `as const` construct, it provides additional information to the Angular compiler required for proper compilation and is a recommended practice (as it makes the exported array immutable from the TypeScript point of view).
```

*   **module injector** with providers configured in `@NgModule.providers` or `@Injectable({providedIn: "..."})`. Those application-wide providers are visible to all components in as well as to other services configured in a module injector.
*   **node injectors** configured in `@Directive.providers` / `@Component.providers` or `@Component.viewProviders`. Those providers are visible to a given component and all its children only.

#### Environment injectors

Making `NgModule`s optional will require new ways of configuring "module" injectors with application-wide providers (for example, [HttpClient](https://angular.io/api/common/http/HttpClient)). In the standalone application (one created with `bootstrapApplication`), “module” providers can be configured during the bootstrap process, in the `providers` option: 

```ts
bootstrapApplication(PhotoAppComponent, {
  providers: [
    {provide: BACKEND_URL, useValue: 'https://photoapp.looknongmodules.com/api'},
    {provide: PhotosService, useClass: PhotosService},
    // ...
  ]
});
```

> The new bootstrap API gives us back the means of configuring “module injectors” without using `NgModule`s.  In this sense, the “module” part of the name is no longer relevant and we’ve decided to introduce a new term: “environment injectors”. 

Reviewer 3: ```suggestion
The new bootstrap API gives us back the means of configuring “module injectors” without using `NgModule`s. In this sense, the “module” part of the name is no longer relevant and we’ve decided to introduce a new term: “environment injectors”. 
```

Environment injectors can be configured using one of the following:

*   `@NgModule.providers `(in applications bootstrapping through an `NgModule`);
*   `@Injectable({provideIn: "..."})`(in both the NgModule-based as well as “standalone” applications);
*   `providers` option in the `bootstrapApplication` call (in fully “standalone” applications);
*   `providers` field in a `Route` configuration.

> Angular v14 introduces a new TypeScript type `EnvironmentInjector` to represent this new naming. The  accompanying `createEnvironmentInjector` API that makes it possible to create environment injectors programmatically: 

Reviewer 3: ```suggestion
Angular v14 introduces a new TypeScript type `EnvironmentInjector` to represent this new naming. The accompanying `createEnvironmentInjector` API makes it possible to create environment injectors programmatically: 
```

```ts
import {createEnvironmentInjector} from '@angular/core';

const parentInjector = … // existing environment injector
const childInjector = createEnvironmentInjector([{provide: PhotosService, useClass: CustomPhotosService}], parentInjector);
```

> Environment injectors have one additional capability: they can execute initialization logic when an environment injector gets created (similar to the `NgModule` constructors that get executed when a module injector is created):

Reviewer 3: ```suggestion
Environment injectors have one additional capability, they can execute initialization logic when an environment injector gets created (similar to the `NgModule` constructors that get executed when a module injector is created):
```

```ts
import {createEnvironmentInjector, ENVIRONMENT_INITIALIZER} from '@angular/core';

createEnvironmentInjector([
{provide: PhotosService, useClass: CustomPhotosService},
{provide: ENVIRONMENT_INITIALIZER, useValue: () => {
        console.log("This function runs when this EnvironmentInjector gets created");
}}
]);
```

#### Standalone injectors

In reality, the dependency injectors hierarchy is slightly more elaborate in applications using standalone components. Let’s consider the following example:

```ts
// an existing "datepicker" component with an NgModule
@Component({
        selector: 'datepicker',
        template: '...',
})
class DatePickerComponent {
  constructor(private calendar: CalendarService) {}
}

@NgModule({
        declarations: [DatePickerComponent],
        exports: [DatePickerComponent]
        providers: [CalendarService],
})
class DatePickerModule {
}

@Component({
        selector: 'date-modal',
        template: '<datepicker></datepicker>',
        standalone: true,
        imports: [DatePickerModule]
})
class DateModalComponent {
}
```

> In the above example, the component `DateModalComponent` is standalone - it can be consumed directly and has no NgModule which needs to be imported in order to use it. However, `DateModalComponent` has a dependency, the `DatePickerComponent,` which is imported via its NgModule (the `DatePickerModule`). This NgModule may declare providers (in this case: `CalendarService`) which are required for the `DatePickerComponent` to function correctly.

Reviewer 3: ```suggestion
In the above example, the component `DateModalComponent` is standalone - it can be consumed directly and has no `NgModule` which needs to be imported in order to use it. However, `DateModalComponent` has a dependency, the `DatePickerComponent,` which is imported via its `NgModule` (the `DatePickerModule`). This `NgModule` may declare providers (in this case: `CalendarService`) which are required for the `DatePickerComponent` to function correctly.
```

> When Angular creates a standalone component, it needs to know that the current injector has all of the necessary services for the standalone component's dependencies, including those based on NgModules. To guarantee that, in some cases Angular will create a new "standalone injector" as a child of the current environment injector. Today, this happens for all bootstrapped standalone components: it will be a child of the root environment injector. The same rule applies to the dynamically created (for example, by the router or the `ViewContainerRef` API) standalone components. 

Reviewer 3: ```suggestion
When Angular creates a standalone component, it needs to know that the current injector has all of the necessary services for the standalone component's dependencies, including those based on `NgModules`. To guarantee that, in some cases Angular will create a new "standalone injector" as a child of the current environment injector. Today, this happens for all bootstrapped standalone components: it will be a child of the root environment injector. The same rule applies to the dynamically created (for example, by the router or the `ViewContainerRef` API) standalone components. 
```
